File: src/vodoo/activities.py
"""Odoo activity (``mail.activity``) operations for Vodoo."""

from __future__ import annotations

from typing import Any

from vodoo.exceptions import RecordOperationError
from vodoo.generated.activities import GeneratedActivityNamespace


class ActivityNamespace(GeneratedActivityNamespace):
    """Namespace for ``mail.activity`` operations."""

    def done(self, activity_id: int) -> Any:
        """Mark an active activity as done and return Odoo's action result."""
        activity = self.get(activity_id, fields=["active"])
        if not activity.get("active"):
            msg = f"Activity {activity_id} is already done"
            raise RecordOperationError(msg)
        return self._client.execute(self._model, "action_done", [activity_id])


def build_activity_domain(
    *,
    model: str | None = None,
    user: str | None = None,
    activity_type: str | None = None,
) -> list[Any]:
    """Build a domain for activity list filtering."""
    domain: list[Any] = []
    if model:
        domain.append(("res_model", "=", model))
    if user:
        domain.append(("user_id.name", "ilike", user))
    if activity_type:
        domain.append(("activity_type_id.name", "ilike", activity_type))
    return domain


__all__ = ["ActivityNamespace", "build_activity_domain"]

File: packages/typescript/src/namespaces/activities.ts
import type { OdooClientApi } from "../client-api.js";
import { RecordOperationError } from "../errors.js";
import { GeneratedActivityNamespace } from "../generated/activities.js";
import type { Domain, OdooRecord } from "../types.js";
import type { ListOptions } from "./domain.js";

export interface ActivityRecord extends OdooRecord {
  id: number;
  date_deadline?: Date | null;
  create_date?: Date | null;
  write_date?: Date | null;
}

export interface ActivityDomainOptions {
  model?: string;
  user?: string;
  activityType?: string;
}

export function buildActivityDomain(
  options: ActivityDomainOptions = {},
): Domain {
  const domain: Array<Domain[number]> = [];
  if (options.model !== undefined && options.model !== "") {
    domain.push(["res_model", "=", options.model]);
  }
  if (options.user !== undefined && options.user !== "") {
    domain.push(["user_id.name", "ilike", options.user]);
  }
  if (options.activityType !== undefined && options.activityType !== "") {
    domain.push(["activity_type_id.name", "ilike", options.activityType]);
  }
  return domain;
}

export class ActivityNamespace extends GeneratedActivityNamespace {
  constructor(client: OdooClientApi) {
    super(client);
  }

  override async list(options: ListOptions = {}): Promise<ActivityRecord[]> {
    return (await super.list(options)) as ActivityRecord[];
  }

  override async get(
    recordId: number,
    fields?: readonly string[],
  ): Promise<ActivityRecord> {
    return (await super.get(recordId, fields)) as ActivityRecord;
  }

  async done(activityId: number): Promise<unknown> {
    const activity = await this.get(activityId, ["active"]);
    if (!activity.active) {
      throw new RecordOperationError(`Activity ${activityId} is already done`);
    }
    return this.client.execute("mail.activity", "action_done", [[activityId]]);
  }
}
